x/evm/keeper/state_transition.go

```diff
@@ -218,31 +218,43 @@ func (k *Keeper) ApplyUnsignedMessage(ctx sdk.Context, txHash common.Hash, msg e
 			res.Logs = types.NewLogsFromEth(receipt.Logs)
 			ctx.EventManager().EmitEvents(tmpCtx.EventManager().Events())
 		}
+	} else {
+		receipt.Status = ethtypes.ReceiptStatusFailed
 	}
 
 	// k.RefundGas doesn't work in omni's context, because we don't actually run the ante handler for each of these tx
 	// so the fee colelctor doesn't get the fees in the first place, and therefore doesn't have enough to refund what is due
 	// refund gas in order to match the Ethereum gas consumption instead of the default SDK one.
 	// if err = k.RefundGas(ctx, msg, msg.Gas()-res.GasUsed, cfg.Params.EvmDenom); err != nil {
-	// 	return nil, errorsmod.Wrapf(err, "failed to refund gas leftover gas to sender %s", msg.From())
+	// 	return nil, nil, errorsmod.Wrapf(err, "failed to refund gas leftover gas to sender %s", msg.From())
 	// }
 
 	if len(receipt.Logs) > 0 {
 		// Update transient block bloom filter
 		k.SetBlockBloomTransient(ctx, receipt.Bloom.Big())
 		k.SetLogSizeTransient(ctx, uint64(txConfig.LogIndex)+uint64(len(receipt.Logs)))
+	} else {
+		// logs cannot be empty
+		receipt.Logs = []*ethtypes.Log{}
 	}
 
 	k.SetTxIndexTransient(ctx, uint64(txConfig.TxIndex)+1)
 
 	totalGasUsed, err := k.AddTransientGasUsed(ctx, res.GasUsed)
 	if err != nil {
-		return nil, errorsmod.Wrap(err, "failed to add transient gas used")
+		return nil, nil, errorsmod.Wrap(err, "failed to add transient gas used")
 	}
 
 	// reset the gas meter for current cosmos transaction
 	k.ResetGasMeterAndConsumeGas(ctx, totalGasUsed)
-	return res, nil
+
+	k.Logger(ctx).Debug("[omni-debug] receipt while executing xchain tx", "receipt", receipt)
+
+	receiptMarshal, err := receipt.MarshalJSON()
```

Comment:
receiptMarshal name doesnt show this is json
why dont we return receipt from this method, and let whoever calls it marshal it to json?

Reply:
This function is a custom function we added to ethermint so this is only called in context of omni xchain txs so I did not think there's a need to shift the responsibility to the caller

Reply:
updated